### PR TITLE
Bug Fixes and Fuzzer Updates

### PR DESCRIPTION
**Pull request by ennamarie19**
This PR fixes two bugs discovered by OSS-Fuzz and updates fuzzer to get past blockers to increase coverage.

### CODE REVIEW

**Comment by jacadzaca** [comment]
hey, very cool PR! Shouldn't we be adding the test cases that the fuzzer generated to our normal test suite? If someone accidentally deletes the fixes introduced by this PR regular tests won't show it.

**Comment by niccokunzmann** [comment]
Yes, I like the PR. There is not a requirement to have tests for improvements but tests covering the change definitely help us in the future. 
I would like to know: Where did you find the bugs, what does it look like?

File: src/icalendar/prop.py [hunk 1]
```diff
@@ -250,7 +252,7 @@ def __init__(self, dt_list):
         self.dts = vDDD
 
     def to_ical(self):
-        dts_ical = (dt.to_ical() for dt in self.dts)
+        dts_ical = (from_unicode(dt.to_ical()) for dt in self.dts)
```
Comment (niccokunzmann) [inline src/icalendar/prop.py:255]: Hi, I do not block a merge, just want to have some conversation: How did you discover that this is helpful? What problem does it solve?

File: src/icalendar/parser_tools.py [hunk 2]
```diff
@@ -2,6 +2,23 @@
 DEFAULT_ENCODING = 'utf-8'
 
 
+def from_unicode(value, encoding='utf-8') -> bytes:
```
Comment (niccokunzmann) [inline src/icalendar/parser_tools.py]: This seems to be a useful function. What are the inputs and what is the data it works on?

File: CHANGES.rst [hunk 3]
```diff
@@ -19,6 +19,9 @@ New features:
 Bug fixes:
 
 - ...
+- Fixed index error in cal.py when attempting to pop from an empty stack
+- Fixed type error in prop.py when attempting to join strings into a byte-string
+- Caught Wrong Date Format in ical_fuzzer to resolve fuzzing coverage blocker
```
Comment (niccokunzmann) [inline CHANGES.rst:24]: I think, if you split the PR into PRs for fixes that are independent and another one for the fuzzer code, some might be merged faster because it is very clear what it is about.

**Comment by niccokunzmann** [comment]
To me it seems that the PR covers some edge cases and improves the fuzzer code. I am sure it improves the project! I am curious about how the fuzzing works and access to the errors and such. I will start a discussion.

**Comment by niccokunzmann** [comment]
I think that personally, I have a bit of a tension in me when I think about submitting code that is not tested by the test cases but by an external entity (Google's OSS FUZZ) that does not allow the same level of access to the test results. (GitHub actions are already a bit in-transparent as you need an account to see the test runs). The code changes are not big and they can be tested. I am happy to merge the changes to the fuzzer code. Also, if we are able to understand the fuzzer code a bit better, then we could probably fuzz a but more - maybe use the example files we have. I am curious to understand more bit by bit so I can add that to other projects, too.

TL DR: For me merging PRs, I am fine with having Google's OSS FUZZ as an informant but not as a dependency.

E.g. looking at the stack error with the additional 'END': Can that be reproduced? Probably. The fuzzer will find it again when the lines are removed, I guess. But I would like our tests to become red once that line is changed again or removed so that I can test offline on my machine and I know what is going on, without Google's mercy. It is a bit political, not technical as it also touches on accessibility and centralization of power. My goal is not to have a political discussion about the PR but to name that that is a question in the background when changes flow from information sources out of my reach. I have no problem with trusting people. I am only careful when it comes to trusting Google's presents. 

I opened #578, so we can discuss the fuzzing and give more people access to the information.

I will generate a test case for the stack.

File: src/icalendar/cal.py [hunk 4]
```diff
@@ -358,6 +358,10 @@ def from_ical(cls, st, multiple=False):
             elif uname == 'END':
                 # we are done adding properties to this component
                 # so pop it from the stack and add it to the new top.
+                if not stack:
```
Comment (niccokunzmann) [inline src/icalendar/cal.py:361]: This is what I would like to create a test case for.

Comment (niccokunzmann) [reply src/icalendar/cal.py:361]: I created https://github.com/ennamarie19/icalendar/pull/2 as a PR to your PR @ennamarie19. This way, the stack pop has a test. I am unsure about the other function, yet. I wonder how the fuzzer found it.

I hope you do not mind that the PR sparks so much text! It is really cool what the fuzzer finds! I am excited and I think @jacadzaca is, too. It just needs a bit of adjustment time to do something new and I am sure we will improve the project a lot!

**Comment by ennamarie19** [comment]
> Yes, I like the PR. There is not a requirement to have tests for improvements but tests covering the change definitely help us in the future. I would like to know: Where did you find the bugs, what does it look like?

I'll keep the tests in mind for future fixes! As for the bugs, they are on the oss-fuzz dashboard for the project as well as sent via an email alert daily. That's what I needed an email address for! :) So I can add you to the distribution list.

File: src/icalendar/prop.py (hunk 1 above)
Comment (ennamarie19) [reply src/icalendar/prop.py:255]: I downloaded the crashing testsuite provided by OSS-Fuzz and passed it into the fuzzer as its input. Then, discovered that the testcase was crashing in this function due to attempting to join a string as bytes.

File: src/icalendar/parser_tools.py (hunk 2 above)
Comment (ennamarie19) [reply src/icalendar/parser_tools.py]: It will accept any data type input but if that input is a string, it will encode it to bytes.

File: CHANGES.rst (hunk 3 above)
Comment (ennamarie19) [reply CHANGES.rst:24]: okay! I can do that!

**Comment by ennamarie19** [comment]
Not at all! I appreciate your enthusiasm and quick responses! Excited to see what comes of it! On Nov 4, 2023, at 2:49 PM, Nicco Kunzmann ***@***.***> wrote:﻿
@niccokunzmann commented on this pull request.



In src/icalendar/cal.py:
> @@ -358,6 +358,10 @@ def from_ical(cls, st, multiple=False):
             elif uname == 'END':
                 # we are done adding properties to this component
                 # so pop it from the stack and add it to the new top.
+                if not stack:

I created ennamarie19#2 as a PR to your PR @ennamarie19. This way, the stack pop has a test. I am unsure about the other function, yet. I wonder how the fuzzer found it.
I hope you do not mind that the PR sparks so much text! It is really cool what the fuzzer finds! I am excited and I think @jacadzaca is, too. It just needs a bit of adjustment time to do something new and I am sure we will improve the project a lot!

File: src/icalendar/prop.py (hunk 1 above)
Comment (niccokunzmann) [reply src/icalendar/prop.py:255]: Oh, that is interesting. Is that something that we can add to the project as tests that other people can run, too?

**Comment by niccokunzmann** [comment]
I am a bit confused, too... Some commits seems useful by themselves but when I have a look at the changes of this PR, the useful change does not occur! I wanted to see if I can cherry-pick certain commits and split it up by myself but I am confused at the moment.

Another confusion: You ([ennamarie19](https://github.com/ennamarie19)) open a PR but all commits are by [capuanob](https://github.com/capuanob). Why is that?

Would it be ok for you if I request the following:
- Open one PR that improves the FUZZ testing of the calendar in this repository (you have write access)
- Later, we create new PRs to that branch to fix the tests
- When the tests run, then we merge the branch

Would that be ok for you?

File: src/icalendar/prop.py (hunk 1 above)
Comment (ennamarie19) [reply src/icalendar/prop.py:255]: Definitely, we can migrate the corpus from OSSFuzz to be hostedi n the `fuzzing` directory

**Comment by ennamarie19** [comment]
> I am a bit confused, too... Some commits seems useful by themselves but when I have a look at the changes of this PR, the useful change does not occur! I wanted to see if I can cherry-pick certain commits and split it up by myself but I am confused at the moment.
> 
> Another confusion: You ([ennamarie19](https://github.com/ennamarie19)) open a PR but all commits are by [capuanob](https://github.com/capuanob). Why is that?
> 
> Would it be ok for you if I request the following:
> 
> * Open one PR that improves the FUZZ testing of the calendar in this repository (you have write access)
> * Later, we create new PRs to that branch to fix the tests
> * When the tests run, then we merge the branch
> 
> Would that be ok for you?

My boyfriend and I work on these fuzz integrations together, so that's why some commits come from him.

I will work on the PR now

**Comment by ennamarie19** [comment]
All changes reflected in new PRs, broken out as requested. Thank you!
